oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/endpoint/JoseHeader.java

```diff
@@ -0,0 +1,391 @@
+import org.springframework.security.oauth2.jose.JwaAlgorithm;
+import org.springframework.security.oauth2.jwt.Jwt;
+import org.springframework.util.Assert;
+
+/*
+ * NOTE:
+ * This originated in gh-9208 (JwtEncoder),
+ * which is required to realize the feature in gh-8175 (JWT Client Authentication).
+ * However, we decided not to merge gh-9208 as part of the 5.5.0 release
+ * and instead packaged it up privately with the gh-8175 feature.
+ * We MAY merge gh-9208 in a later release but that is yet to be determined.
+ *
+ * gh-9208 Introduce JwtEncoder
+ * https://github.com/spring-projects/spring-security/pull/9208
+ *
+ * gh-8175 Support JWT for Client Authentication
+ * https://github.com/spring-projects/spring-security/issues/8175
+ */
+
+/**
+ * The JOSE header is a JSON object representing the header parameters of a JSON Web
+ * Token, whether the JWT is a JWS or JWE, that describe the cryptographic operations
+ * applied to the JWT and optionally, additional properties of the JWT.
+ *
+ * @author Anoop Garlapati
+ * @author Joe Grandja
+ * @since 5.5
+ * @see Jwt
+ * @see <a target="_blank" href="https://tools.ietf.org/html/rfc7519#section-5">JWT JOSE
+ * Header</a>
+ * @see <a target="_blank" href="https://tools.ietf.org/html/rfc7515#section-4">JWS JOSE
+ * Header</a>
+ * @see <a target="_blank" href="https://tools.ietf.org/html/rfc7516#section-4">JWE JOSE
+ * Header</a>
+ */
+final class JoseHeader {
+
+	private final Map<String, Object> headers;
+
+	private JoseHeader(Map<String, Object> headers) {
+		this.headers = Collections.unmodifiableMap(new HashMap<>(headers));
+	}
+
+	/**
+	 * Returns the {@link JwaAlgorithm JWA algorithm} used to digitally sign the JWS or
+	 * encrypt the JWE.
+	 * @return the {@link JwaAlgorithm}
+	 */
+	@SuppressWarnings("unchecked")
+	<T extends JwaAlgorithm> T getAlgorithm() {
+		return (T) getHeader(JoseHeaderNames.ALG);
+	}
+
+	/**
+	 * Returns the JWK Set URL that refers to the resource of a set of JSON-encoded public
+	 * keys, one of which corresponds to the key used to digitally sign the JWS or encrypt
+	 * the JWE.
+	 * @return the JWK Set URL
+	 */
+	URL getJwkSetUri() {
+		return getHeader(JoseHeaderNames.JKU);
+	}
+
+	/**
+	 * Returns the JSON Web Key which is the public key that corresponds to the key used
+	 * to digitally sign the JWS or encrypt the JWE.
+	 * @return the JSON Web Key
+	 */
+	Map<String, Object> getJwk() {
+		return getHeader(JoseHeaderNames.JWK);
+	}
+
+	/**
+	 * Returns the key ID that is a hint indicating which key was used to secure the JWS
+	 * or JWE.
+	 * @return the key ID
+	 */
+	String getKeyId() {
+		return getHeader(JoseHeaderNames.KID);
+	}
+
+	/**
+	 * Returns the X.509 URL that refers to the resource for the X.509 public key
+	 * certificate or certificate chain corresponding to the key used to digitally sign
+	 * the JWS or encrypt the JWE.
+	 * @return the X.509 URL
+	 */
+	URL getX509Uri() {
```

Comment:
The spec states:

> The "x5u" (X.509 URL) Header Parameter is a URI [RFC3986] that refers
   to a resource

It is a `URI` hence the suffix in the name. It returns a `URL` given it points to a a resource.

Reply:
It also says `JWK Set URL`. A URL is a specific type of URI. Unless there are cases where the this would only be a URI (doesn't appear that way), I think we should change this to be `.getJwkSetUrl` since it is more precise.